base/reducedim.jl

```diff
@@ -124,8 +118,33 @@ function _reducedim_init(f, op, fv, fop, A, region)
     return reducedim_initarray(A, region, z, Tr)
 end
 
-reducedim_init(f, op::typeof(max), A::AbstractArray{T}, region) where {T} = reducedim_initarray0(A, region, f, maximum)
-reducedim_init(f, op::typeof(min), A::AbstractArray{T}, region) where {T} = reducedim_initarray0(A, region, f, minimum)
+# initialization when computing minima and maxima requires a little care
+for (f1, f2, initval) in ((:min, :max, :Inf), (:max, :min, :(-Inf)))
+    @eval function reducedim_init(f, op::typeof($f1), A::AbstractArray, region)
+        # First compute the reduce indices. This will throw an ArgumentError
+        # if any region is invalid
+        ri = reduced_indices(A, region)
+
+        # Next, throw if reduction is over a region with length zero
+        any(i -> isempty(axes(A, i)), region) && _empty_reduce_error()
+
+        # Make a view of the first slice of the region
+        A1 = view(A, ri...)
+
+        if isempty(A1)
+            # If the slice is empty just return non-view version as the initial array
+            return copy(A1)
+        else
+            # otherwise use the min/max of the first slice as initial value
+            v0 = mapreduce(f, $f2, A1)
+
+            # but NaNs need to be avoided as intial values
+            v0 = v0 != v0 ? typeof(v0)($initval) : v0
```

Comment:
AFAICT the `mapreduce` call above will give `NaN` whenever one of the values is `NaN`. So this code is going to use the fallback `initval` even if there are non-`NaN` values which could be used. Maybe it would be better to use a custom loop to go over the first slice and skip `NaN` values?

Actually, maybe something similar to [the `mapfirst!` method](https://github.com/JuliaLang/julia/pull/28027/files#diff-c4a56019a8dae547ca5556ad8df4d1f0R358) I had to write at https://github.com/JuliaLang/julia/pull/27845 to skip missing values is needed.

Anyway, a few tests for these corner cases would be good. ;-)

Reply:
This is basically a check for floats but tries to be a little more generic. Using the `initval` is fine for floats and only floats will probably ever be `NaN` so I don't think there is an issue here.